doc/port_auto_neg/port-auto-negotiation-design.md

```diff
@@ -0,0 +1,500 @@
+
+This command always replace the advertised speeds instead of append. For example, say the current advertised speeds value are "10000,25000", if user configure it with `config interface advertised-speeds Ethernet0 40000,100000`, the advertised speeds value will be changed to "40000,100000".
+
+##### Config interface type
+
+Configuring interface type takes effect only if auto negotiation is disabled. If auto negotiation is enabled, this command still saves interface type value to CONFIG_DB.
+
+```
+Format:
+  config interface type <interface_name> <interface_type>
+
+Arguments:
+  interface_name: name of the interface to be configured. e.g: Ethernet0
+  interface_type: interface type, valid value include: KR4, SR4 and so on. A list of valid interface type could be found at saiport.h.
+
+Example:
+  config interface type Ethernet0 KR4
+
+Return:
+  error message if interface_name or interface_type is invalid otherwise empty
+```
+
+##### Config advertised interface types
+
+Configuring advertised interface types takes effect only if auto negotiation is enabled. If auto negotiation is disabled, this command still saves advertised interface types value to CONFIG_DB.
+
+```
+Format:
+  config interface advertised-types <interface_name> <interface_type_list>
+
+Arguments:
+  interface_name: name of the interface to be configured. e.g: Ethernet0
+  interface_type_list: a list of interface types to be advertised or "all". e.g: KR4,SR4.
+
+Example:
+  config interface advertised-types Ethernet0 KR4,SR4
+  config interface advertised-types all
+
+Return:
+  error message if interface_name or interface_type_list is invalid otherwise empty
+
+Note:
+  interface_type_list value "all" means all supported interface type 
+```
+
+This command always replace the advertised interface types instead of append. For example, say the current advertised interface types value are "KR4,SR4", if user configure it with `config interface advertised-types Ethernet0 CR4`, the advertised interface types value will be changed to "CR4".
+
+##### Show interfaces auto negotiation status
+
+As command `show interfaces status` already has 11 columns, a new CLI command will be added to display the port auto negotiation status. All data of this command are fetched from **APPL_DB**.
+
+```
+Format:
+  show interfaces auto-neg-status <interface_name>
+
+Arguments:
+  interface_name: optional. Name of the interface to be shown. e.g: Ethernet0. If interface_name is not given, this command shows auto negotiation status for all interfaces.
+
+Example:
+  show interfaces auto-neg-status
+  show interfaces auto-neg-status Ethernet0
+
+Return:
+  error message if interface_name is invalid otherwise:
+
+  Interface    Auto-Neg Mode    Speed    Adv Speeds    Type    Adv Types
+-----------  ---------------  -------  ------------  ------  -----------
+  Ethernet0          enabled     100G      40G,100G     N/A      CR4,KR4
+ Ethernet32         disabled      40G           N/A     N/A          N/A
```

Comment:
can we show the cable interface type here?

Reply:
Hi @lguohan , could you please elaborate more about the "cable interface type" here? I guess maybe you are talking about the "type" column in the output of `show interfaces status`. And its value is usually something like "QSFP28 or later" or "QSFP+ or later". Or is it something in the SFP eeprom data:

```
Ethernet100: SFP EEPROM detected
        Application Advertisement: N/A
        Connector: No separable connector
        Encoding: 64B66B
        Extended Identifier: Power Class 3(2.5W max), CDR present in Rx Tx
        Extended RateSelect Compliance: QSFP+ Rate Select Version 1
        Identifier: QSFP28 or later
        Length Cable Assembly(m): 3
        Nominal Bit Rate(100Mbs): 255
        Specification compliance:
                Extended Specification compliance: 100G AOC (Active Optical Cable) or 25GAUI C2M AOC
        Vendor Date Code(YYYY-MM-DD Lot): 2019-02-15
        Vendor Name: Mellanox
        Vendor OUI: 00-02-c9
        Vendor PN: MFA1A00-C003
        Vendor Rev: B2
        Vendor SN: MT1907FT08074
```

Thanks.